raytracing/tests/testsAsphericInterface.py

```diff
@@ -0,0 +1,178 @@
+        n2 = asphere.surfaceNormal(y=-abs(R/2))
+        self.assertTrue(n2 < 0)
+
+    def testSurfaceNormalWithKappa(self):
+
+        R = 10
+        for kappa in [0, -0.5, -1, -1.5, -100]:
+            asphere = AsphericInterface(R=R, kappa=kappa, n1=1, n2=1.5)
+
+            self.assertAlmostEqual(asphere.surfaceNormal(y=0), 0)
+
+            for y in [-2*R, -R, -R/2, 0, R/2, R, 2*R]:
+                if asphere.z(y) is not None:
+                    # Then angle must be defined
+                    n1 = asphere.surfaceNormal(y=y)
+                    self.assertIsNotNone(n1)
+
+                    n2 = asphere.surfaceNormal(y=-y)
+                    self.assertIsNotNone(n2)
+
+                    self.assertAlmostEqual(n1, -n2)
+                else:
+                    self.assertIsNone(asphere.surfaceNormal(y=y))
+
+
+    def testNormalIncidenceRefraction(self):
+        asphere = AsphericInterface(R=10, kappa=0, n1=1, n2=1.5)
+        paraxialRay = Ray(y=0, theta=0)
+        self.assertAlmostEqual(asphere.surfaceNormal(y=paraxialRay.y),0)
+
+        paraxialOut = asphere.mul_ray_paraxial(paraxialRay)
+        nonparaxialOut = asphere.mul_ray_nonparaxial(paraxialRay)
+        self.assertAlmostEqual(paraxialRay.y, nonparaxialOut.y)
+        self.assertAlmostEqual(paraxialRay.theta, nonparaxialOut.theta)
+
+    def testOnAxisUpwardIncidenceRefraction(self):
+        asphere = AsphericInterface(R=10, kappa=0, n1=1, n2=1.5)
+        paraxialRay = Ray(y=0, theta=0.1)
+        self.assertAlmostEqual(asphere.surfaceNormal(y=paraxialRay.y),0)
+
+        paraxialOut = asphere.mul_ray_paraxial(paraxialRay)
+        nonparaxialOut = asphere.mul_ray_nonparaxial(paraxialRay)
+        self.assertAlmostEqual(paraxialOut.y, nonparaxialOut.y, 3)
+        self.assertAlmostEqual(paraxialOut.theta, nonparaxialOut.theta, 3)
+
+    def testOffAxisUpwardIncidenceRefraction(self):
+        asphere = AsphericInterface(R=10, kappa=0, n1=1, n2=1.5)
+        paraxialRay = Ray(y=0.1, theta=0.1)
+
+        paraxialOut = asphere.mul_ray_paraxial(paraxialRay)
+        nonparaxialOut = asphere.mul_ray_nonparaxial(paraxialRay)
+        self.assertAlmostEqual(paraxialOut.y, nonparaxialOut.y, 3)
+        self.assertAlmostEqual(paraxialOut.theta, nonparaxialOut.theta, 3)
+
+    def testCollection(self):
+        asphere = AsphericInterface(R=10, kappa=0, n1=1, n2=1.5)
+        rays =[Ray(0, 0),Ray(0, 0.1),Ray(0.1, 0),Ray(0.1, 0.1),
+        Ray(0.1, -0.1), Ray(-0.1, 0.1),Ray(0, -0.1),Ray(-0.1, 0),
+        Ray(-0.1, -0.1)] 
+
+        for paraxialRay in rays:
+            paraxialOut = asphere.mul_ray_paraxial(paraxialRay)
+            nonparaxialOut = asphere.mul_ray_nonparaxial(paraxialRay)
+            self.assertAlmostEqual(paraxialOut.y, nonparaxialOut.y, 3)
+            self.assertAlmostEqual(paraxialOut.theta, nonparaxialOut.theta, 3)
+
+    @envtest.skip
+    def testOpticalPath(self):
+        Matrix.useNonParaxialCalculations = True
+        path = ImagingPath()
```

Comment:
Maybe have `ImagingPath(useNonParaxialCalculations=None)` instead ? Just for usage. And I guess we keep the global in Matrix.

Reply:
It should be in OpticalPath or something like that for the user, but deep down, it is at the matrix level that this affects the calculation. You are right that it could go in the ImagingPath/OpticalPath `__init__`.

See the above comment that I made essentially 2 minutes before you made the present comment: I am not sure anymore if it should be "a global choice", because I am not sure the results are correct.